pkg/controller/infrastructure/actuator.go

```diff
@@ -332,54 +344,113 @@ func (a *actuator) shareCustomizedImages(ctx context.Context, infra *extensionsv
+	a.logger.Info("Share customized image with Shoot's Alicloud account from Seed", "infrastructure", infra.Name)
 	for _, worker := range cluster.Shoot.Spec.Provider.Workers {
-		imageID, err := helper.FindImageForRegionFromCloudProfile(cloudProfileConfig, worker.Machine.Image.Name, *worker.Machine.Image.Version, infra.Spec.Region)
-		if err != nil {
-			if providerStatus := infra.Status.ProviderStatus; providerStatus != nil {
-				infrastructureStatus := &apisalicloud.InfrastructureStatus{}
-				if _, _, err := a.Decoder().Decode(providerStatus.Raw, nil, infrastructureStatus); err != nil {
-					return nil, errors.Wrapf(err, "could not decode infrastructure status of infrastructure '%s'", kutil.ObjectName(infra))
-				}
-
-				machineImage, err := helper.FindMachineImage(infrastructureStatus.MachineImages, worker.Machine.Image.Name, *worker.Machine.Image.Version)
-				if err != nil {
-					return nil, err
-				}
-				imageID = machineImage.ID
-			} else {
+		var machineImage *apisalicloud.MachineImage
+		if common.UseEncryptedSystemDisk(worker.Volume) {
+			if machineImage, err = a.ensureEncryptedImageForShootProviderAccount(ctx, cloudProfileConfig, worker, infra, shootAlicloudROSClient); err != nil {
+				return nil, err
+			}
+		} else {
+			if machineImage, err = a.ensurePlainImageForShootProviderAccount(ctx, cloudProfileConfig, worker, infra, shootAlicloudECSClient, shootCloudProviderAccountID); err != nil {
 				return nil, err
 			}
 		}
-		machineImages = appendMachineImage(machineImages, alicloudv1alpha1.MachineImage{
-			Name:    worker.Machine.Image.Name,
-			Version: *worker.Machine.Image.Version,
-			ID:      imageID,
-		})
 
-		// if this is a whitelisted machine image, we no longer need to check if it exists in cloud provider account, and
-		// we don't need to share the image to that account either.
-		if a.isWhitelistedImageID(imageID) {
-			a.logger.Info("Skipping image sharing for whitelisted image", "imageID", imageID)
-			continue
+		machineImages = helper.AppendMachineImage(machineImages, *machineImage)
+	}
+
+	return machineImages, nil
+}
+
+func (a *actuator) ensureEncryptedImageForShootProviderAccount(ctx format.Ctx, cloudProfileConfig *apisalicloud.CloudProfileConfig, worker gardenerv1beta1.Worker, infra *extensionsv1alpha1.Infrastructure, shootROSClient alicloudclient.ROS) (*apisalicloud.MachineImage, error) {
+	infrastructureStatus := &apisalicloud.InfrastructureStatus{}
+	if infra.Status.ProviderStatus != nil {
+		if _, _, err := a.Decoder().Decode(infra.Status.ProviderStatus.Raw, nil, infrastructureStatus); err != nil {
+			return nil, errors.Wrapf(err, "could not decode infrastructure status of infrastructure '%s'", kutil.ObjectName(infra))
 		}
+	}
+	if machineImage, err := helper.FindMachineImage(infrastructureStatus.MachineImages, worker.Machine.Image.Name, *worker.Machine.Image.Version, true); err == nil {
+		return machineImage, nil
+	}
```

Comment:
What about machine image update scenario? Because it will always check whether current machine image used in status is encrypted. If it is, then the one specified in cloud profile will no longer be encrypted.

Reply:
The image status stored encrypted flag and when we find an available image, encrypted flag is always considered.
For update scenario, if original one is plain and will switch to use encrypted one, the extension will help create an encrypted image for shoot. Image specified in cloud profile is always plain.

Reply:
Oh, okay. My mistake, version is also checked when finding a machine image here.

Reply:
Ah okay. My mistake, the version is also checked here.